cirq/work/observable_settings.py

```diff
@@ -0,0 +1,186 @@
+# This file contains code modified from PyQuil.
+# https://github.com/rigetti/pyquil
+# The original code's license is:
+#
+# Copyright 2016-2019 Rigetti Computing
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Modifications are covered under:
+#
+# Copyright 2020 The Cirq developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dataclasses
+from typing import Union, Iterable, Dict, TYPE_CHECKING, Tuple
+
+from cirq import ops, value, protocols
+
+if TYPE_CHECKING:
+    import cirq
+    from cirq.value.product_state import _NamedOneQubitState
+
+
+@dataclasses.dataclass(frozen=True)
+class InitObsSetting:
+    """A pair of initial state and observable.
+
+    Usually, given a circuit you want to iterate through many
+    InitObsSettings to vary the initial state preparation and output
+    observable.
```

Comment:
This is true, but it doesn't necessarily justify a class to bundle the two together. What are the use-cases for it?

One use-case I found below is to specify top of the iteration range. This can be easily accomplished by specifying top of the range of initial states and top of the range of observables.

Reply:
The entire premise of this feature is you want to measure many input-output pairs efficiently. So these need to be tuple-fied for a) term grouping #3372  b) circuit pre- and post- rotation generation

Reply:
Why not use the existing python type `Tuple[ProductState, PauliString]` instead of defining a new class? The only significant logic present here is the validation and I imagine we may be able to fold that into a validation step elsewhere since we're still going to need to ensure that the initial state and observable sizes are consistent with the circuit.

Reply:
validation logic, nice string representation, being able to address the components by name makes the rest of the code more readable, dataclass means low overhead to give a name to this tuple-like object

Reply:
Fair enough :-)